blazesym-go/symbolizer.go

```diff
@@ -0,0 +1,121 @@
+package blazesym
+
+import (
+	"unsafe"
+)
+
+/*
+#cgo LDFLAGS: -lblazesym_c
+#include "blazesym.h"
+
+// Adding a C function to return syms from blaze_result
+struct blaze_sym* get_result(blaze_syms* res, size_t pos) {
+	return &res->syms[pos];
+}
+*/
+import "C"
+
+// Symbolizer represents a Blazesym symbolizer.
+type Symbolizer struct {
+	s *C.blaze_symbolizer
+}
+
+// NewSymbolizer creates an instance of a symbolizer.
+// See: https://docs.rs/blazesym-c/latest/blazesym_c/fn.blaze_symbolizer_new.html
+func NewSymbolizer() (*Symbolizer, error) {
+	s := C.blaze_symbolizer_new()
+	if s == nil {
+		return nil, BlazeErr(C.blaze_err_last()).Error()
+	}
+
+	return &Symbolizer{s: s}, nil
+}
+
+// Close closes the a symbolizer.
+func (s *Symbolizer) Close() {
+	C.blaze_symbolizer_free(s.s)
+}
+
+// SymbolizeElfVirtOffsets symbolizes virtual offsets in an ELF file.
+// See: https://docs.rs/blazesym-c/latest/blazesym_c/fn.blaze_symbolize_elf_virt_offsets.html
+func (s *Symbolizer) SymbolizeElfVirtOffsets(path string, debugSyms bool, stack []uint64) ([]Sym, error) {
+	symSrcElf := C.struct_blaze_symbolize_src_elf{}
+	symSrcElf.type_size = C.ulong(unsafe.Sizeof(symSrcElf))
+	symSrcElf.path = C.CString(path)
+	symSrcElf.debug_syms = C.bool(debugSyms)
+	defer C.free(unsafe.Pointer(symSrcElf.path))
+
+	caddr, clen := stackToPtr(stack)
+
+	return s.processSyms(C.blaze_symbolize_elf_virt_offsets(s.s, &symSrcElf, caddr, clen), stack)
+}
+
+// SymbolizeProcessAbsAddrs symbolizes a list of process absolute addresses.
+// See: https://docs.rs/blazesym-c/latest/blazesym_c/fn.blaze_symbolize_process_abs_addrs.html
+func (s *Symbolizer) SymbolizeProcessAbsAddrs(pid uint32, debugSyms bool, noMapFiles bool, noVdso bool, stack []uint64) ([]Sym, error) {
+	symSrcProcess := C.struct_blaze_symbolize_src_process{}
+	symSrcProcess.type_size = C.ulong(unsafe.Sizeof(symSrcProcess))
+	symSrcProcess.pid = C.uint32_t(pid)
+	symSrcProcess.debug_syms = C.bool(debugSyms)
+	symSrcProcess.no_map_files = C.bool(noMapFiles)
+	symSrcProcess.no_vdso = C.bool(noVdso)
+
+	caddr, clen := stackToPtr(stack)
+
+	return s.processSyms(C.blaze_symbolize_process_abs_addrs(s.s, &symSrcProcess, caddr, clen), stack)
+}
+
+func (s *Symbolizer) processSyms(syms *C.blaze_syms, stack []uint64) ([]Sym, error) {
+	lastErr := BlazeErr(C.blaze_err_last())
```

Comment:
This pattern of checking last error here seems dangerous, as there is no guarantee that `syms` hasn't been stored anywhere and other `blazesym` calls have happened in between. I'd rather we move this check to the call site.